### PR TITLE
License Fixes

### PR DESCRIPTION
**Pull request by gbiellem**
Related to #782 #783 and #789

Added a feature to periodically poll for license changes and expose the license details via rest api 

Changes the Management utility to block install and upgrading instances if the license has expired or the upgrade version falls outside the maintenance period

### CODE REVIEW

File: src/ServiceControl/Licensing/LicenseModule.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        public ActiveLicense ActiveLicense { get; set; }
+
+        public LicenseModule()
+        {
+            Get["/license"] = parameters =>
+            {
+                var licenseInfo = new LicenseInfo
+                {
+                    TrialLicense = ActiveLicense.Details.IsTrialLicense,
+                    Edition = ActiveLicense.Details.Edition ?? "" ,
```
Comment (johnsimons) [inline src/ServiceControl/Licensing/LicenseModule.cs]: `String.Empty` ?

File: src/ServiceControl/Infrastructure/Nancy/Modules/RootModule.cs [hunk 2]
```diff
@@ -30,6 +30,7 @@ public RootModule()
                     Name = SettingsReader<string>.Read("Name", "ServiceControl"),
                     Description = SettingsReader<string>.Read("Description", "The management backend for the Particular Service Platform"),
                     LicenseStatus = License.IsValid ? "valid" : "invalid",
```
Comment (johnsimons) [inline src/ServiceControl/Infrastructure/Nancy/Modules/RootModule.cs]: Do we move the status to `LicenseDetails` ?

Comment (gbiellem) [inline src/ServiceControl/Infrastructure/Nancy/Modules/RootModule.cs]: Wouldn;t that break older versions of pulse?

File: src/ServiceControl/Licensing/LicenseModule.cs (hunk 1 above)
Comment (gbiellem) [inline src/ServiceControl/Licensing/LicenseModule.cs]: fixed

**Comment by gbiellem** [comment]
@johnsimons  are you happy with the rest of this?

File: src/ServiceControl/Licensing/LicenseCheckFeature.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+                manualResetEvent.WaitOne();
+            }
+        }
+
+        void Refresh(object _)
+        {
+            activeLicense.Refresh();
+            try
+            {
+                timer.Change((int)TimeSpan.FromMinutes(10).TotalMilliseconds, -1);
```
Comment (johnsimons) [inline src/ServiceControl/Licensing/LicenseCheckFeature.cs]: Does it need to be this frequent ?

**Comment by johnsimons** [comment]
i thought we also wanted to warn about license about to expire ?

File: src/ServiceControl/Licensing/LicenseCheckFeature.cs (hunk 3 above)
Comment (gbiellem) [inline src/ServiceControl/Licensing/LicenseCheckFeature.cs]: How frequent do you suggest? Having it too long would be a weird experience for anyone updating the license..  i.e.  I updated the license but ServicePulse still says it's expired.  We already get those support cases

Comment (johnsimons) [inline src/ServiceControl/Licensing/LicenseCheckFeature.cs]: 1 hour

**Comment by gbiellem** [comment]
> i thought we also wanted to warn about license about to expire ?

I'm not sure I see a benefit in adding a warning x days before expiry.  Pulse will tell them when the license has expired and there is no degradation of service once it has expired.  I'm been trying to think of another product that does a impending license expiry and I can't come up with one.   

Given we are now exposing the license expiration and maintenance we could also do that from within pulse without using a notification

File: src/ServiceControl/Licensing/LicenseCheckFeature.cs (hunk 3 above)
Comment (johnsimons) [inline src/ServiceControl/Licensing/LicenseCheckFeature.cs]: BTW, you can change that to use `TimeKeeper`, see https://github.com/Particular/ServiceControl/blob/40607aacd062404253797a1f3221718d2177598c/src/ServiceControl/Recoverability/Retries/FailedMessageRetries.cs#L124

Comment (gbiellem) [inline src/ServiceControl/Licensing/LicenseCheckFeature.cs]: I just had a look at the Raven code - it checks for a license every 15 minutes.  An hour seems like a long time.  If we're going to set it to be that long then I'll change the mgmt UI to indicate that when updating/adding a license

Comment (gbiellem) [inline src/ServiceControl/Licensing/LicenseCheckFeature.cs]: Thanks for the TimeKeeper tip - Will update

**Comment by gbiellem** [comment]
@johnsimons Changed to Timekeeper and set the license refresh to 8 hours.   Also added a query string option to force the license to reload.  Tweaked mgmt util to try to call the license refresh on each instance after license is imported so we don;t need a high refresh rate

**Comment by johnsimons** [comment]
looks good @gbiellem

**Comment by gbiellem** [comment]
One last thing to do.  The license expirty only blocks if installs and upgrades  are done through the mgmt UI.   Once I add the checks to the Powershell cmdlets this should be good to go

**Comment by gbiellem** [comment]
@johnsimons When I tested the artifact from the build service I'm getting some binding errors in the management UI that I don't get when debugging.  Only difference i can see is this branch is 4.5 but after merging with master it's 4.5.2.  So don't merge yet til I resolve this

**Comment by gbiellem** [comment]
@johnsimons  Think I'm done with this. Appreciate if you could take another look and merge if you are happy with it.

File: src/ServiceControl.Config/Commands/AddInstanceCommand.cs [hunk 4]
```diff
@@ -5,21 +5,31 @@
...
             this.addInstance = addInstance;
+            this.installer = installer;
         }
 
         public override void Execute(object obj)
         {
+            var licenseCheckResult = installer.CheckLicenseIsValid();
+            if (!licenseCheckResult.Valid)
+            {
+                windowManager.ShowMessage(licenseCheckResult.Message, "Installation could not continue due to an issue with the current license. Contact sales@particular.net", hideCancel: true);
```
Comment (johnsimons) [inline src/ServiceControl.Config/Commands/AddInstanceCommand.cs]: @gbiellem, it seems the error message `licenseCheckResult.Message` is actually displayed as the title of the popup dialog, shouldn't the `licenseCheckResult.Message` be part of the body ?

File: src/ServiceControl.Config/Commands/UpgradeInstanceCommand.cs [hunk 5]
```diff
@@ -28,13 +28,19 @@ public UpgradeInstanceCommand(Func<InstanceDetailsViewModel, bool> canExecuteMet
...
             this.eventAggregator = eventAggregator;
             this.installer = installer;
         }
 
         public override async Task ExecuteAsync(InstanceDetailsViewModel model)
         {
+            var licenseCheckResult = installer.CheckLicenseIsValid();
+            if (!licenseCheckResult.Valid)
+            {
+                windowManager.ShowMessage(licenseCheckResult.Message, "Upgrade could not continue due to an issue with the current license. Contact sales@particular.net", hideCancel: true);
```
Comment (johnsimons) [inline src/ServiceControl.Config/Commands/UpgradeInstanceCommand.cs]: @gbiellem, it seems the error message `licenseCheckResult.Message` is actually displayed as the title of the popup dialog, shouldn't the `licenseCheckResult.Message` be part of the body ?

File: src/ServiceControlInstaller.Engine/FileSystem/DeploymentZipInfo.cs [hunk 6]
```diff
@@ -55,5 +57,58 @@ public void ValidateZip()
...
+        public bool TryReadServiceControlReleaseDate(out DateTime releaseDate)
+        {
+            releaseDate = DateTime.MinValue;
+            var tempDomain = AppDomain.CreateDomain("TemporaryAppDomain");
+            var tempFile = Path.Combine(Path.GetTempPath(), @"ServiceControl\ServiceControl.exe");
+            try
+            {
+                using (var zip = ZipFile.Read(FilePath))
+                {
+                    var entry = zip.Entries.FirstOrDefault(p => p.FileName == "ServiceControl/ServiceControl.exe");
```
Comment (johnsimons) [inline src/ServiceControlInstaller.Engine/FileSystem/DeploymentZipInfo.cs]: is the slash right ?

File: src/ServiceControlInstaller.Engine/Unattended/UnattendInstaller.cs [hunk 7]
```diff
@@ -228,6 +243,45 @@ public bool Delete(string InstanceName, bool removeDB, bool removeLogs)
...
+            if (!license.Details.ValidForServiceControl)
+            {
+                return new CheckLicenseResult(false, "This license edition does not include ServiceControl");
+            }
+
+            if (ZipInfo.TryReadServiceControlReleaseDate(out releaseDate))
+            {
+                if (license.Details.ReleaseNotCoveredByMaintenance(releaseDate))
+                {
+                    return new CheckLicenseResult(false, "This version was released afer the upgrade protection expiration date specified in the license");
```
Comment (johnsimons) [inline src/ServiceControlInstaller.Engine/Unattended/UnattendInstaller.cs]: spelling "afer"

File: src/ServiceControlInstaller.Engine/FileSystem/DeploymentZipInfo.cs (hunk 6 above)
Comment (gbiellem) [inline src/ServiceControlInstaller.Engine/FileSystem/DeploymentZipInfo.cs]: Yep - Slash is correct. Entries in a zip use that slash

File: src/ServiceControl.Config/Commands/AddInstanceCommand.cs (hunk 4 above)
Comment (gbiellem) [inline src/ServiceControl.Config/Commands/AddInstanceCommand.cs]: that was intentional.

File: src/ServiceControl.Config/Commands/UpgradeInstanceCommand.cs (hunk 5 above)
Comment (gbiellem) [inline src/ServiceControl.Config/Commands/UpgradeInstanceCommand.cs]: that was intentional.

File: src/ServiceControlInstaller.Engine/Unattended/UnattendInstaller.cs (hunk 7 above)
Comment (gbiellem) [inline src/ServiceControlInstaller.Engine/Unattended/UnattendInstaller.cs]: oops - will fix

File: src/ServiceControl.Config/Commands/UpgradeInstanceCommand.cs (hunk 5 above)
Comment (johnsimons) [inline src/ServiceControl.Config/Commands/UpgradeInstanceCommand.cs]: not following ? Doesn't it look strange that way ?

**Comment by gbiellem** [comment]
@johnsimons I've addressed your comments above. Anything else or we good to merge?

**Pull request merged**